netbench-orchestrator/src/russula/mod.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{task::Poll, time::Duration};
+use paste::paste;
+use std::{collections::BTreeSet, net::SocketAddr};
+use tokio::net::TcpStream;
+use tracing::{error, info};
+
+mod error;
+mod event;
+mod network_utils;
+mod protocol;
+mod states;
+
+use error::{RussulaError, RussulaResult};
+use protocol::Protocol;
+
+struct ProtocolInstance<P: Protocol> {
+    pub addr: SocketAddr,
+    pub stream: TcpStream,
+    pub protocol: P,
+}
+
+/// A coordination framework.
+///
+/// Russula is a wrapper over a pair of Coordinator and Worker [Protocol]s.
+/// A Coordinator can be used to synchronize multiple workers across
+/// different hosts.
+pub struct Russula<P: Protocol> {
+    /// List of protocol instance to synchronize with.
+    instance_list: Vec<ProtocolInstance<P>>,
+
+    /// Polling frequency when trying to make progress.
+    poll_delay: Duration,
+}
+
+macro_rules! state_api {
+    {
+        $(#[$meta:meta])*
+            $state:ident
+    } => {paste!{
+
+    $(#[$meta])*
+    pub async fn [<run_till_ $state>](&mut self) -> RussulaResult<()> {
+        while self.[<poll_ $state>]().await?.is_pending() {
+            tokio::time::sleep(self.poll_delay).await;
+        }
+
+        Ok(())
+    }
+
+    $(#[$meta])*
+    pub async fn [<poll_ $state>](&mut self) -> RussulaResult<Poll<()>> {
+        // Poll each peer protocol instance.
+        //
+        // If the peer is already in the desired state then this should be a noop.
+        for peer in self.instance_list.iter_mut() {
+            if let Err(err) = peer.protocol.[<poll_ $state>](&mut peer.stream).await {
+                if err.is_fatal() {
+                    error!("{} {}", err, peer.addr);
+                    panic!("{} {}", err, peer.addr);
+                }
+            }
+        }
+
+        // Check that all instances are at the desired state.
+        let poll = if self.[<is_ $state _state>]() {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        };
+        Ok(poll)
+    }
+
+    /// Check if all instances are at the desired state
+    fn [< is_ $state _state>](&self) -> bool {
+        for peer in self.instance_list.iter() {
+            // All instance must be at the desired state
+            if !peer.protocol.[< is_ $state _state>]() {
+                return false;
+            }
+        }
+        true
+    }
+}};
+}
+
+impl<P: Protocol + Send> Russula<P> {
```

Comment:
I wonder if you could avoid the macros here by using `enums` for the states instead?

E.g. instead of
```
is_ready_state()
is_done_state()
is_woker_running_state()
```

You could just have
```
is_state(ProtocolState::Ready)
```
Or something like that.

Reply:
yea macros can be annoying to read. lemme try and see how using enum looks like